src/audio/selector/selector_generic.c

```diff
@@ -171,171 +171,145 @@ static void sel_s32le_nch(struct comp_dev *dev, struct audio_stream __sparse_cac
 #else
 #if CONFIG_FORMAT_S16LE
 /**
- * \brief Channel selection for 16 bit, 1 channel data format.
- * \param[in,out] mod Selector base module device.
+ * \brief Mixing routine for 16-bit, m channel input x n channel output single frame.
+ * \param[out] dst Sink buffer.
+ * \param[in] dst_channels Number of sink channels.
+ * \param[in] src Source data.
+ * \param[in] src_channels Number of source channels.
+ * \param[in] coeffs_config IPC4 micsel config with coefficients.
+ */
+static void process_frame_s16le(int16_t dst[], int dst_channels,
+				int16_t src[], int src_channels,
+				struct ipc4_selector_coeffs_config *coeffs_config)
+{
+	int32_t accum;
+	int i;
+	int j;
+
+	for (i = 0; i < dst_channels; i++) {
+		accum = 0;
+		for (j = 0; j < src_channels; j++)
+			accum += (int32_t)src[j] * (int32_t)coeffs_config->coeffs[i][j];
+		dst[i] = (int16_t)(accum >> 10);
+	}
```

Comment:
we recently added a matrix operation library, but this multiplication isn't there, right? @singalsu and this Q10 (?) multiplication is too specific to be added there?

Reply:
In fact, the matrix.c library looks useful for this 16-bit x 16-bit multiplication, but unfortunately this wouldn't be the case for 32-bit x 16-bit multiplication in subsequent 'process_frame_s32le' function. The proposed approach was created for the sake of consistency.

Reply:
Feel free to add 32 x 16 version if you like the approach. I hope we can optimize in near future the libraries with intrinsics and then all clients will get the benefit.

Reply:
I believe we can switch to matrix approach in the future, when 32x16 version gets ready. Hope this does not prevent this PR from being merged now.